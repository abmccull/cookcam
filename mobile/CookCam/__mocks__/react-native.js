const React = require('react');
const RN = jest.requireActual('react-native');

// Mock PixelRatio for StyleSheet
const PixelRatio = {
  roundToNearestPixel: (value) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size) => size * 2,
};

// Override modules that cause issues in tests
module.exports = {
  ...RN,
  PixelRatio,
  StyleSheet: {
    ...RN.StyleSheet,
    create: (styles) => styles,
    flatten: (style) => {
      if (!style) return {};
      if (Array.isArray(style)) {
        return Object.assign({}, ...style.filter(s => s));
      }
      return style;
    },
    absoluteFill: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    absoluteFillObject: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    hairlineWidth: 1,
  },
  Platform: {
    ...RN.Platform,
    OS: 'ios',
    select: (obj) => obj.ios || obj.default,
    Version: 14,
    isTV: false,
    isTesting: true,
  },
  Dimensions: {
    get: (dim) => {
      if (dim === 'window') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      if (dim === 'screen') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      return { width: 375, height: 812 };
    },
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    set: jest.fn(),
  },
  Animated: {
    ...RN.Animated,
    timing: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    spring: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    Value: class {
      constructor(value) {
        this._value = value;
      }
      setValue(value) {
        this._value = value;
      }
      interpolate(config) {
        return this;
      }
      addListener(callback) {
        return '1';
      }
      removeListener(id) {}
      removeAllListeners() {}
      stopAnimation(callback) {
        callback && callback(this._value);
      }
      resetAnimation(callback) {
        callback && callback(this._value);
      }
      setOffset(offset) {}
      flattenOffset() {}
      extractOffset() {}
    },
    View: RN.View,
    Text: RN.Text,
    Image: RN.Image,
    ScrollView: RN.ScrollView,
    createAnimatedComponent: (Component) => Component,
    parallel: (animations, config) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    sequence: (animations) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    loop: (animation) => ({
      start: (callback) => {
        animation.start && animation.start();
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    event: () => jest.fn(),
    add: jest.fn(),
    subtract: jest.fn(),
    divide: jest.fn(),
    multiply: jest.fn(),
    modulo: jest.fn(),
    diffClamp: jest.fn(),
  },
  NativeModules: {
    ...RN.NativeModules,
    UIManager: {
      RCTView: {
        directEventTypes: {},
      },
    },
    PlatformConstants: {
      forceTouchAvailable: false,
    },
    RNCNetInfo: {
      getCurrentState: jest.fn(() => Promise.resolve({
        isConnected: true,
        isInternetReachable: true,
        type: 'wifi'
      })),
      addListener: jest.fn(),
      removeListeners: jest.fn(),
    },
    StatusBarManager: {
      HEIGHT: 20,
      getHeight: jest.fn(),
      setColor: jest.fn(),
      setStyle: jest.fn(),
      setHidden: jest.fn(),
      setNetworkActivityIndicatorVisible: jest.fn(),
      setBackgroundColor: jest.fn(),
      setTranslucent: jest.fn(),
    },
    SettingsManager: {
      settings: {},
      setValues: jest.fn(),
      deleteValues: jest.fn(),
    },
    DeviceInfo: {
      Dimensions: {
        window: { width: 375, height: 812, scale: 2, fontScale: 1 },
        screen: { width: 375, height: 812, scale: 2, fontScale: 1 },
      },
    },
  },
  TouchableOpacity: React.forwardRef((props, ref) =>
    React.createElement('TouchableOpacity', { ...props, ref })
  ),
  TouchableHighlight: React.forwardRef((props, ref) =>
    React.createElement('TouchableHighlight', { ...props, ref })
  ),
  TouchableWithoutFeedback: React.forwardRef((props, ref) =>
    React.createElement('TouchableWithoutFeedback', { ...props, ref })
  ),
  TouchableNativeFeedback: {
    SelectableBackground: jest.fn(),
    SelectableBackgroundBorderless: jest.fn(),
    Ripple: jest.fn(),
    canUseNativeForeground: jest.fn(),
  },
  View: React.forwardRef((props, ref) =>
    React.createElement('View', { ...props, ref })
  ),
  Text: React.forwardRef((props, ref) =>
    React.createElement('Text', { ...props, ref })
  ),
  Image: React.forwardRef((props, ref) =>
    React.createElement('Image', { ...props, ref })
  ),
  ScrollView: React.forwardRef((props, ref) =>
    React.createElement('ScrollView', { ...props, ref })
  ),
  TextInput: React.forwardRef((props, ref) =>
    React.createElement('TextInput', { ...props, ref })
  ),
  Modal: React.forwardRef((props, ref) =>
    React.createElement('Modal', { ...props, ref })
  ),
  FlatList: React.forwardRef((props, ref) =>
    React.createElement('FlatList', { ...props, ref })
  ),
  SectionList: React.forwardRef((props, ref) =>
    React.createElement('SectionList', { ...props, ref })
  ),
  VirtualizedList: React.forwardRef((props, ref) =>
    React.createElement('VirtualizedList', { ...props, ref })
  ),
  Alert: {
    alert: jest.fn(),
    prompt: jest.fn(),
  },
  Keyboard: {
    addListener: jest.fn(() => ({ remove: jest.fn() })),
    removeListener: jest.fn(),
    dismiss: jest.fn(),
    scheduleLayoutAnimation: jest.fn(),
  },
  Linking: {
    openURL: jest.fn(() => Promise.resolve()),
    canOpenURL: jest.fn(() => Promise.resolve(true)),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AppState: {
    currentState: 'active',
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AccessibilityInfo: {
    isScreenReaderEnabled: jest.fn(() => Promise.resolve(false)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    announceForAccessibility: jest.fn(),
    setAccessibilityFocus: jest.fn(),
  },
  findNodeHandle: jest.fn(() => 1),
  unstable_batchedUpdates: (callback) => callback(),
};