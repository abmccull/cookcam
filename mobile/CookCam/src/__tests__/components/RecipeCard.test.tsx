import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react-native';

// Set up global PixelRatio before React Native imports it
global.PixelRatio = {
  roundToNearestPixel: (value: number) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size: number) => size * 2,
};

// Mock React Native modules BEFORE importing the component
jest.mock('react-native', () => {
  const actualRN = jest.requireActual('react-native');
  // Provide PixelRatio to StyleSheet
  actualRN.PixelRatio = global.PixelRatio;
  
  return {
    ...actualRN,
    PixelRatio: global.PixelRatio,
    StyleSheet: {
      ...actualRN.StyleSheet,
      create: (styles: any) => styles,
      flatten: (style: any) => style,
    },
  };
});

// Mock environment config
jest.mock('../../config/env', () => ({
  __esModule: true,
  default: () => ({
    SUPABASE_URL: "https://test.supabase.co",
    SUPABASE_ANON_KEY: "test-anon-key",
    API_BASE_URL: "https://test-api.cookcam.com",
  }),
}));

// Mock services
jest.mock('../../services/cookCamApi', () => ({
  cookCamApi: {
    getRecipeNutrition: jest.fn(() => 
      Promise.resolve({
        success: true,
        data: {
          calories: 450,
          protein: 25,
          carbs: 45,
          fat: 18,
          fiber: 5,
          sodium: 680,
        }
      })
    ),
  },
}));

// Mock utils
jest.mock('../../utils/logger', () => ({
  __esModule: true,
  default: {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  },
}));

jest.mock('../../utils/responsive', () => ({
  scale: (n: number) => n,
  verticalScale: (n: number) => n,
  moderateScale: (n: number) => n,
}));

// Mock styles
jest.mock('../../styles', () => ({
  tokens: {
    colors: {
      primary: '#4CAF50',
      text: {
        primary: '#000000',
        secondary: '#666666',
        light: '#999999',
        inverse: '#FFFFFF',
      },
      background: {
        main: '#FFFFFF',
        secondary: '#F5F5F5',
        accent: '#E8F5E9',
        overlay: 'rgba(0,0,0,0.3)',
        overlayLight: 'rgba(0,0,0,0.5)',
      },
      difficulty: {
        easy: '#4CAF50',
        medium: '#FF9800',
        hard: '#F44336',
      },
      interactive: {
        favorite: '#E91E63',
      },
      brand: {
        chef: '#FF6B35',
      },
      border: {
        primary: '#E0E0E0',
      },
    },
    spacing: {
      xs: 4,
      sm: 12,
      md: 16,
    },
    fontSize: {
      sm: 12,
      base: 14,
      lg: 18,
    },
    fontWeight: {
      medium: '500',
      semibold: '600',
      bold: '700',
    },
    borderRadius: {
      medium: 8,
      large: 12,
    },
  },
  mixins: {
    cards: {
      base: {
        backgroundColor: '#FFFFFF',
        borderRadius: 12,
        overflow: 'hidden',
      },
    },
    layout: {
      flex1: { flex: 1 },
      flexRow: { flexDirection: 'row' },
      centerContent: {
        justifyContent: 'center',
        alignItems: 'center',
      },
      centerHorizontal: {
        alignItems: 'center',
      },
      absoluteTopLeft: {
        position: 'absolute',
        top: 0,
        left: 0,
      },
      absoluteBottomRight: {
        position: 'absolute',
        bottom: 0,
        right: 0,
      },
    },
    avatars: {
      small: {
        width: 24,
        height: 24,
        borderRadius: 12,
      },
    },
  },
  styleUtils: {
    combine: (...styles: any[]) => Object.assign({}, ...styles),
  },
}));

// Mock icons
jest.mock('lucide-react-native', () => ({
  Heart: 'Heart',
  Clock: 'Clock',
  Flame: 'Flame',
  MessageCircle: 'MessageCircle',
  Share2: 'Share2',
}));

// Mock components
jest.mock('../../components/ChefBadge', () => {
  const mockReact = require('react');
  return function MockChefBadge({ tier, size }: any) {
    return mockReact.createElement('MockChefBadge', { 
      testID: `ChefBadge-${tier}-${size}` 
    }, `ChefBadge-${tier}-${size}`);
  };
});

jest.mock('../../components/NutritionBadge', () => {
  const mockReact = require('react');
  return function MockNutritionBadge({ nutrition, servings, variant }: any) {
    return mockReact.createElement('MockNutritionBadge', { 
      testID: `NutritionBadge-${variant}` 
    }, `NutritionBadge-${variant}`);
  };
});

// Now import the component after all mocks are set up
import RecipeCard from '../../components/RecipeCard';
import { cookCamApi } from '../../services/cookCamApi';

describe('RecipeCard', () => {
  const mockRecipe = {
    id: '1',
    title: 'Spaghetti Carbonara',
    image: 'https://example.com/carbonara.jpg',
    cookTime: 30,
    difficulty: 'Medium',
    likes: 125,
    comments: 42,
    servings: 4,
    creator: {
      name: 'Chef Mario',
      tier: 3 as const,
      avatar: 'https://example.com/chef.jpg',
    },
  };

  const mockOnPress = jest.fn();
  const mockOnLike = jest.fn();
  const mockOnComment = jest.fn();
  const mockOnShare = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render correctly with all props', () => {
      const { toJSON } = render(
        <RecipeCard
          recipe={mockRecipe}
          onPress={mockOnPress}
          onLike={mockOnLike}
          onComment={mockOnComment}
          onShare={mockOnShare}
          showNutrition={true}
        />
      );
      
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render recipe title', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('Spaghetti Carbonara')).toBeTruthy();
    });

    it('should render cook time with icon', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('30 min')).toBeTruthy();
      expect(screen.UNSAFE_getByType('Clock')).toBeTruthy();
    });

    it('should render difficulty badge', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('Medium')).toBeTruthy();
    });

    it('should render creator name', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('Chef Mario')).toBeTruthy();
    });

    it('should render creator initial in avatar', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('C')).toBeTruthy();
    });

    it('should render chef badge when tier is provided', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByTestId('ChefBadge-3-small')).toBeTruthy();
    });

    it('should not render chef badge when tier is not provided', () => {
      const recipeWithoutTier = {
        ...mockRecipe,
        creator: { ...mockRecipe.creator, tier: undefined },
      };
      render(<RecipeCard recipe={recipeWithoutTier} />);
      
      expect(screen.queryByText(/ChefBadge/)).toBeFalsy();
    });

    it('should render action buttons with counts', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('125')).toBeTruthy(); // likes
      expect(screen.getByText('42')).toBeTruthy(); // comments
      expect(screen.getByText('Share')).toBeTruthy();
    });

    it('should render action icons', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.UNSAFE_getByType('Heart')).toBeTruthy();
      expect(screen.UNSAFE_getByType('MessageCircle')).toBeTruthy();
      expect(screen.UNSAFE_getByType('Share2')).toBeTruthy();
    });

    it('should render flame icon as placeholder', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.UNSAFE_getByType('Flame')).toBeTruthy();
    });
  });

  describe('Interactions', () => {
    it('should call onPress when card is pressed', () => {
      render(<RecipeCard recipe={mockRecipe} onPress={mockOnPress} />);
      
      const card = screen.UNSAFE_getAllByType('TouchableOpacity')[0];
      fireEvent.press(card);
      
      expect(mockOnPress).toHaveBeenCalledTimes(1);
    });

    it('should call onLike when like button is pressed', () => {
      render(<RecipeCard recipe={mockRecipe} onLike={mockOnLike} />);
      
      const likeButton = screen.UNSAFE_getAllByType('TouchableOpacity')[1];
      fireEvent.press(likeButton);
      
      expect(mockOnLike).toHaveBeenCalledTimes(1);
    });

    it('should call onComment when comment button is pressed', () => {
      render(<RecipeCard recipe={mockRecipe} onComment={mockOnComment} />);
      
      const commentButton = screen.UNSAFE_getAllByType('TouchableOpacity')[2];
      fireEvent.press(commentButton);
      
      expect(mockOnComment).toHaveBeenCalledTimes(1);
    });

    it('should call onShare when share button is pressed', () => {
      render(<RecipeCard recipe={mockRecipe} onShare={mockOnShare} />);
      
      const shareButton = screen.UNSAFE_getAllByType('TouchableOpacity')[3];
      fireEvent.press(shareButton);
      
      expect(mockOnShare).toHaveBeenCalledTimes(1);
    });

    it('should not call handlers when they are not provided', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      // Test that component renders without crashing when no handlers provided
      // The mocks ensure handlers are not called
      expect(mockOnPress).not.toHaveBeenCalled();
      expect(mockOnLike).not.toHaveBeenCalled();
      expect(mockOnComment).not.toHaveBeenCalled();
      expect(mockOnShare).not.toHaveBeenCalled();
    });
  });

  describe('Nutrition Data', () => {
    it('should fetch nutrition data when showNutrition is true', async () => {
      render(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledWith('1');
      });
    });

    it('should display nutrition badges after fetching', async () => {
      render(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(screen.getByTestId('NutritionBadge-compact')).toBeTruthy();
        expect(screen.getByTestId('NutritionBadge-full')).toBeTruthy();
      });
    });

    it('should not fetch nutrition data when showNutrition is false', async () => {
      render(<RecipeCard recipe={mockRecipe} showNutrition={false} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).not.toHaveBeenCalled();
      }, { timeout: 100 });
    });

    it('should handle nutrition fetch error gracefully', async () => {
      (cookCamApi.getRecipeNutrition as jest.Mock).mockRejectedValueOnce(
        new Error('Network error')
      );
      
      render(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalled();
      });
      
      expect(screen.queryByText(/NutritionBadge/)).toBeFalsy();
    });

    it('should handle empty nutrition response', async () => {
      (cookCamApi.getRecipeNutrition as jest.Mock).mockResolvedValueOnce({
        success: false,
        data: null,
      });
      
      render(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalled();
      });
      
      expect(screen.queryByText(/NutritionBadge/)).toBeFalsy();
    });

    it('should not show nutrition badges while loading', () => {
      render(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      // Before async fetch completes
      expect(screen.queryByText(/NutritionBadge/)).toBeFalsy();
    });
  });

  describe('Difficulty Styling', () => {
    it('should apply correct styling for easy difficulty', () => {
      const easyRecipe = { ...mockRecipe, difficulty: 'Easy' };
      render(<RecipeCard recipe={easyRecipe} />);
      
      expect(screen.getByText('Easy')).toBeTruthy();
    });

    it('should apply correct styling for medium difficulty', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      expect(screen.getByText('Medium')).toBeTruthy();
    });

    it('should apply correct styling for hard difficulty', () => {
      const hardRecipe = { ...mockRecipe, difficulty: 'Hard' };
      render(<RecipeCard recipe={hardRecipe} />);
      
      expect(screen.getByText('Hard')).toBeTruthy();
    });

    it('should handle unknown difficulty levels', () => {
      const unknownRecipe = { ...mockRecipe, difficulty: 'Expert' };
      render(<RecipeCard recipe={unknownRecipe} />);
      
      expect(screen.getByText('Expert')).toBeTruthy();
    });

    it('should handle lowercase difficulty values', () => {
      const lowerRecipe = { ...mockRecipe, difficulty: 'easy' };
      render(<RecipeCard recipe={lowerRecipe} />);
      
      expect(screen.getByText('easy')).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle missing creator name', () => {
      const recipeWithEmptyName = {
        ...mockRecipe,
        creator: { ...mockRecipe.creator, name: '' },
      };
      render(<RecipeCard recipe={recipeWithEmptyName} />);
      
      expect(screen.getByText('?')).toBeTruthy(); // Should show ? for initial
    });

    it('should handle single character names', () => {
      const singleCharRecipe = {
        ...mockRecipe,
        creator: { ...mockRecipe.creator, name: 'A' },
      };
      render(<RecipeCard recipe={singleCharRecipe} />);
      
      expect(screen.getAllByText('A')).toHaveLength(2); // Initial and name
    });

    it('should handle long recipe titles', () => {
      const longTitleRecipe = {
        ...mockRecipe,
        title: 'This is a very long recipe title that should be truncated properly in the UI to maintain good visual layout and not break the design',
      };
      render(<RecipeCard recipe={longTitleRecipe} />);
      
      const titleElement = screen.getByText(longTitleRecipe.title);
      expect(titleElement.props.numberOfLines).toBe(2);
    });

    it('should handle zero likes and comments', () => {
      const noEngagementRecipe = {
        ...mockRecipe,
        likes: 0,
        comments: 0,
      };
      render(<RecipeCard recipe={noEngagementRecipe} />);
      
      expect(screen.getAllByText('0')).toHaveLength(2);
    });

    it('should handle missing optional props', () => {
      const minimalRecipe = {
        id: '2',
        title: 'Simple Recipe',
        cookTime: 15,
        difficulty: 'Easy',
        likes: 10,
        comments: 5,
        creator: {
          name: 'Home Cook',
        },
      };
      
      const { toJSON } = render(<RecipeCard recipe={minimalRecipe} />);
      expect(toJSON()).toBeTruthy();
    });

    it('should handle very large numbers', () => {
      const popularRecipe = {
        ...mockRecipe,
        likes: 999999,
        comments: 123456,
        cookTime: 1440, // 24 hours
      };
      render(<RecipeCard recipe={popularRecipe} />);
      
      expect(screen.getByText('999999')).toBeTruthy();
      expect(screen.getByText('123456')).toBeTruthy();
      expect(screen.getByText('1440 min')).toBeTruthy();
    });
  });

  describe('Performance', () => {
    it('should use React.memo for optimization', () => {
      expect(RecipeCard.displayName).toBe('RecipeCard');
    });

    it('should only fetch nutrition once per recipe id', async () => {
      const { rerender } = render(
        <RecipeCard recipe={mockRecipe} showNutrition={true} />
      );
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledTimes(1);
      });
      
      // Clear mock to track new calls
      (cookCamApi.getRecipeNutrition as jest.Mock).mockClear();
      
      // Rerender with same recipe
      rerender(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      // Wait a bit to ensure no additional calls
      await new Promise(resolve => setTimeout(resolve, 100));
      
      // Should not fetch again
      expect(cookCamApi.getRecipeNutrition).not.toHaveBeenCalled();
    });

    it('should fetch nutrition for different recipe id', async () => {
      const { rerender } = render(
        <RecipeCard recipe={mockRecipe} showNutrition={true} />
      );
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledWith('1');
      });
      
      const newRecipe = { ...mockRecipe, id: '2' };
      rerender(<RecipeCard recipe={newRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledWith('2');
      });
    });

    it('should not refetch when toggling showNutrition off and on', async () => {
      const { rerender } = render(
        <RecipeCard recipe={mockRecipe} showNutrition={true} />
      );
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledTimes(1);
      });
      
      // Turn off nutrition
      rerender(<RecipeCard recipe={mockRecipe} showNutrition={false} />);
      
      // Clear mock
      (cookCamApi.getRecipeNutrition as jest.Mock).mockClear();
      
      // Turn on nutrition again
      rerender(<RecipeCard recipe={mockRecipe} showNutrition={true} />);
      
      await waitFor(() => {
        expect(cookCamApi.getRecipeNutrition).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('Accessibility', () => {
    it('should have proper touch target with activeOpacity', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      const mainCard = screen.UNSAFE_getAllByType('TouchableOpacity')[0];
      expect(mainCard.props.activeOpacity).toBe(0.9);
    });

    it('should handle text truncation properly', () => {
      const longContentRecipe = {
        ...mockRecipe,
        title: 'Very long title that needs to be truncated for proper display',
      };
      
      render(<RecipeCard recipe={longContentRecipe} />);
      
      const title = screen.getByText(longContentRecipe.title);
      expect(title.props.numberOfLines).toBe(2);
    });

    it('should have all interactive elements as TouchableOpacity', () => {
      render(
        <RecipeCard 
          recipe={mockRecipe}
          onPress={mockOnPress}
          onLike={mockOnLike}
          onComment={mockOnComment}
          onShare={mockOnShare}
        />
      );
      
      const touchables = screen.UNSAFE_getAllByType('TouchableOpacity');
      expect(touchables.length).toBe(4); // card + 3 action buttons
    });
  });

  describe('Creator Info', () => {
    it('should display creator avatar with correct initial', () => {
      render(<RecipeCard recipe={mockRecipe} />);
      
      const initial = screen.getByText('C');
      expect(initial).toBeTruthy();
    });

    it('should handle non-latin characters in creator name', () => {
      const intlRecipe = {
        ...mockRecipe,
        creator: { ...mockRecipe.creator, name: '李明' },
      };
      render(<RecipeCard recipe={intlRecipe} />);
      
      expect(screen.getByText('李')).toBeTruthy();
      expect(screen.getByText('李明')).toBeTruthy();
    });

    it('should show tier badge for different tier levels', () => {
      const tierLevels = [1, 2, 3, 4, 5] as const;
      
      tierLevels.forEach(tier => {
        const { unmount } = render(
          <RecipeCard 
            recipe={{
              ...mockRecipe,
              creator: { ...mockRecipe.creator, tier }
            }} 
          />
        );
        
        expect(screen.getByTestId(`ChefBadge-${tier}-small`)).toBeTruthy();
        unmount();
      });
    });
  });
});