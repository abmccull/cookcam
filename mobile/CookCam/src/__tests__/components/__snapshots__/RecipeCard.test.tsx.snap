// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard Rendering should render correctly with all props 1`] = `
<TouchableOpacity
  activeOpacity={0.9}
  onPress={[Function]}
  ref={null}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 12,
      "marginBottom": 16,
      "overflow": "hidden",
    }
  }
>
  <View
    ref={null}
    style={
      {
        "borderTopLeftRadius": 12,
        "borderTopRightRadius": 12,
        "height": 200,
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <View
      ref={null}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E8F5E9",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Flame
        color="#FF6B35"
        size={48}
      />
    </View>
    <View
      ref={null}
      style={
        {
          "backgroundColor": "rgba(0,0,0,0.3)",
          "left": 0,
          "paddingBottom": 24,
          "paddingHorizontal": 12,
          "paddingTop": 12,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        ref={null}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <View
          ref={null}
          style={
            {
              "borderRadius": 12,
              "height": 24,
              "width": 24,
            }
          }
        >
          <Text
            ref={null}
            style={
              {
                "color": "#FFFFFF",
                "fontSize": 14,
                "fontWeight": "700",
              }
            }
          >
            C
          </Text>
        </View>
        <Text
          ref={null}
          style={
            {
              "color": "#FFFFFF",
              "flex": 1,
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Chef Mario
        </Text>
        <MockChefBadge
          testID="ChefBadge-3-small"
        >
          ChefBadge-3-small
        </MockChefBadge>
      </View>
    </View>
    <View
      ref={null}
      style={
        {
          "bottom": 0,
          "flexDirection": "row",
          "gap": 8,
          "margin": 12,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View
        ref={null}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0.5)",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 4,
            "paddingHorizontal": 12,
            "paddingVertical": 4,
          }
        }
      >
        <Clock
          color="#FFFFFF"
          size={14}
        />
        <Text
          ref={null}
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 12,
              "fontWeight": "600",
            }
          }
        >
          30
           min
        </Text>
      </View>
      <View
        ref={null}
        style={
          {
            "backgroundColor": "#FF9800",
            "borderRadius": 8,
            "paddingHorizontal": 12,
            "paddingVertical": 4,
          }
        }
      >
        <Text
          ref={null}
          style={
            {
              "color": "#FFFFFF",
              "fontSize": 12,
              "fontWeight": "600",
            }
          }
        >
          Medium
        </Text>
      </View>
    </View>
  </View>
  <View
    ref={null}
    style={
      {
        "padding": 16,
      }
    }
  >
    <Text
      numberOfLines={2}
      ref={null}
      style={
        {
          "color": "#000000",
          "fontSize": 18,
          "fontWeight": "700",
          "marginBottom": 12,
        }
      }
    >
      Spaghetti Carbonara
    </Text>
    <View
      ref={null}
      style={
        {
          "borderTopColor": "#E0E0E0",
          "borderTopWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": 8,
          "paddingTop": 12,
        }
      }
    >
      <TouchableOpacity
        onPress={[Function]}
        ref={null}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 4,
          }
        }
      >
        <Heart
          color="#E91E63"
          size={20}
        />
        <Text
          ref={null}
          style={
            {
              "color": "#666666",
              "fontSize": 14,
              "fontWeight": "500",
            }
          }
        >
          125
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        onPress={[Function]}
        ref={null}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 4,
          }
        }
      >
        <MessageCircle
          color="#666666"
          size={20}
        />
        <Text
          ref={null}
          style={
            {
              "color": "#666666",
              "fontSize": 14,
              "fontWeight": "500",
            }
          }
        >
          42
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        onPress={[Function]}
        ref={null}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 4,
          }
        }
      >
        <Share2
          color="#666666"
          size={20}
        />
        <Text
          ref={null}
          style={
            {
              "color": "#666666",
              "fontSize": 14,
              "fontWeight": "500",
            }
          }
        >
          Share
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</TouchableOpacity>
`;
